core/src/main/java/com/scalar/db/storage/jdbc/NamespaceMetadataService.java

```diff
@@ -0,0 +1,223 @@
+            + "PRIMARY KEY ("
+            + enclose(COL_NAMESPACE_NAME)
+            + "))";
+    createTable(connection, createTableStatement, true);
+
+    // Insert the system namespace to the namespaces table
+    insertIntoNamespacesTable(connection, metadataSchema);
+  }
+
+  void deleteNamespacesTableIfEmpty(Connection connection) throws SQLException {
+    if (isNamespacesTableEmpty(connection)) {
+      deleteTable(connection, encloseFullTableName(metadataSchema, TABLE_NAME));
+      deleteMetadataSchema(connection);
+    }
+  }
+
+  private boolean isNamespacesTableEmpty(Connection connection) throws SQLException {
+    String selectAllTables = "SELECT * FROM " + encloseFullTableName(metadataSchema, TABLE_NAME);
+
+    Set<String> namespaces = new HashSet<>();
+    try (Statement statement = connection.createStatement();
+        ResultSet results = statement.executeQuery(selectAllTables)) {
+      int count = 0;
+      while (results.next()) {
+        namespaces.add(results.getString(COL_NAMESPACE_NAME));
+        // Only need to fetch the first two rows
+        if (count++ == 2) {
+          break;
+        }
+      }
+    }
+
+    return namespaces.size() == 1 && namespaces.contains(metadataSchema);
+  }
+
+  private void deleteMetadataSchema(Connection connection) throws SQLException {
+    String sql = rdbEngine.deleteMetadataSchemaSql(metadataSchema);
+    execute(connection, sql);
+  }
+
+  void insertIntoNamespacesTable(Connection connection, String namespaceName) throws SQLException {
+    String insertStatement =
+        "INSERT INTO " + encloseFullTableName(metadataSchema, TABLE_NAME) + " VALUES (?)";
+    try (PreparedStatement preparedStatement = connection.prepareStatement(insertStatement)) {
+      preparedStatement.setString(1, namespaceName);
+      preparedStatement.execute();
+    }
+  }
+
+  void upsertIntoNamespacesTable(Connection connection, String namespace) throws SQLException {
+    try {
+      insertIntoNamespacesTable(connection, namespace);
+    } catch (SQLException e) {
+      // ignore if the schema already exists
+      if (!rdbEngine.isDuplicateKeyError(e)) {
+        throw e;
+      }
+    }
+  }
+
+  void deleteFromNamespacesTable(Connection connection, String namespaceName) throws SQLException {
+    String deleteStatement =
+        "DELETE FROM "
+            + encloseFullTableName(metadataSchema, TABLE_NAME)
+            + " WHERE "
+            + enclose(COL_NAMESPACE_NAME)
+            + " = ?";
+    try (PreparedStatement preparedStatement = connection.prepareStatement(deleteStatement)) {
+      preparedStatement.setString(1, namespaceName);
+      preparedStatement.execute();
+    }
+  }
+
+  boolean namespaceExists(Connection connection, String namespace) throws SQLException {
+    String selectQuery =
+        "SELECT 1 FROM "
+            + encloseFullTableName(metadataSchema, TABLE_NAME)
+            + " WHERE "
+            + enclose(COL_NAMESPACE_NAME)
+            + " = ?";
```

Comment:
[minor] It might depend on the underlying RDBMS, but adding `LIMIT 1` might optimize the scan.